pupil_src/shared_modules/fixation_detector.py

```diff
@@ -275,3 +320,464 @@ def get_init_dict(self):
+        }
+
+        self.remove_n_datums(len(self.gaze_data))
+        return new_fixation
+
+    @property
+    def duration(self):
+        if not self.gaze_data: return 0.
+        t0 = self.gaze_data[ 0]['timestamp']
+        t1 = self.gaze_data[-1]['timestamp']
+        return t1-t0
+
+    @property
+    def max_distance(self):
+        distance_count = len(self.distances)
+        if (0 <= self._max_dist_idc_c < distance_count and
+            0 <= self._max_dist_idc_r < distance_count):
+            return self.distances[self._max_dist_idc_c][self._max_dist_idc_r]
+
+        max_dist = 0.
+        for i in range(1, distance_count):
+            for j in range(i):
+                if (max_dist < self.distances[i][j]):
+                    max_dist = self.distances[i][j]
+                    self._max_dist_idc_c = i
+                    self._max_dist_idc_r = j
+        return max_dist
+
+    @property
+    def max_distance_deg(self):
+        return np.rad2deg(self.max_distance)
+
+    def __str__(self):
+        return '<%s %fsec %fdeg>'%(self.__class__.__name__,self.duration,self.max_distance)
+
+class Online_Base_Fixation_Detector(Plugin):
+    order = .2
+    uniqueness = 'by_base_class'
+    def __init__(self, g_pool):
+        super(Online_Base_Fixation_Detector, self).__init__(g_pool)
+
+class Fixation_Detector_3D(Online_Base_Fixation_Detector):
+    """docstring for Online_Fixation_Detector_Pupil_Angle_Dispersion_Duration
+    """
+    def __init__(self, g_pool, max_dispersion=1.0, min_duration=0.15):
+        super(Fixation_Detector_3D, self).__init__(g_pool)
+        self.min_duration = min_duration
+        self.max_dispersion = max_dispersion
+        self.dispersion_slider_min = 0.
+        self.dispersion_slider_max = 3.
+        self.dispersion_slider_stp = 0.05
+        self.detection_windows = [Detection_Window(),Detection_Window()]
+
+    def update(self,frame=None,events={}):
+        recent = [[],[]]
+        events['fixations'] = []
+        for gp in events['gaze_positions']:
+            for base_idx, base_datum in enumerate(gp['base_data']):
+                eye_id = base_datum['id']
+                if 'circle_3d' in base_datum:
+                    recent[eye_id].append((gp, base_datum))
+        for eye_id in range(2):
+            detect_win = self.detection_windows[eye_id]
+            # add recent gaze positions one by one to make sure we do not overstep
+            for datum in recent[eye_id]:
+                detect_win.remove_surplus_data(self.min_duration)
+                detect_win.append(datum)
+                if (detect_win.duration >= self.min_duration and
+                    detect_win.max_distance_deg <= self.max_dispersion):
+                    new_fixation = detect_win.pop_fixation()
+                    events['fixations'].append(new_fixation)
+                    logger.debug('Fixation eye %i: %fdeg'%(new_fixation['eye_id'],new_fixation['dispersion']))

```

Comment:
This is only displayed once a fixation is found. This means this happens at most each `min_duration ` seconds. This will probably be more than 1Hz.